hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java

```diff
@@ -0,0 +1,694 @@
+              " specified in URI but host '" + omHost + "' is a "
+              + "logical (HA) OzoneManager and does not use port information.");
+        }
+      } else {
+        // When port number is not specified, read it from config
+        omPort = OmUtils.getOmRpcPort(conf);
+      }
+
+      SecurityConfig secConfig = new SecurityConfig(conf);
+
+      if (secConfig.isSecurityEnabled()) {
+        this.securityEnabled = true;
+      }
+
+      String replicationTypeConf =
+          conf.get(OzoneConfigKeys.OZONE_REPLICATION_TYPE,
+              OzoneConfigKeys.OZONE_REPLICATION_TYPE_DEFAULT);
+
+      int replicationCountConf = conf.getInt(OzoneConfigKeys.OZONE_REPLICATION,
+          OzoneConfigKeys.OZONE_REPLICATION_DEFAULT);
+
+      if (OmUtils.isOmHAServiceId(conf, omHost)) {
+        // omHost is listed as one of the service ids in the config,
+        // thus we should treat omHost as omServiceId
+        this.ozoneClient =
+            OzoneClientFactory.getRpcClient(omHost, conf);
+      } else if (StringUtils.isNotEmpty(omHost) && omPort != -1) {
+        this.ozoneClient =
+            OzoneClientFactory.getRpcClient(omHost, omPort, conf);
+      } else {
+        this.ozoneClient =
+            OzoneClientFactory.getRpcClient(conf);
+      }
+      objectStore = ozoneClient.getObjectStore();
+      proxy = objectStore.getClientProxy();
+      this.replicationType = ReplicationType.valueOf(replicationTypeConf);
+      this.replicationFactor = ReplicationFactor.valueOf(replicationCountConf);
+      this.configuredDnPort = conf.getInt(
+          OzoneConfigKeys.DFS_CONTAINER_IPC_PORT,
+          OzoneConfigKeys.DFS_CONTAINER_IPC_PORT_DEFAULT);
+    } finally {
+      Thread.currentThread().setContextClassLoader(contextClassLoader);
+    }
+  }
+
+  OzoneBucket getBucket(OFSPath ofsPath, boolean createIfNotExist)
+      throws IOException {
+
+    return getBucket(ofsPath.getVolumeName(), ofsPath.getBucketName(),
+        createIfNotExist);
+  }
+
+  /**
+   * Get OzoneBucket object to operate in.
+   * Optionally create volume and bucket if not found.
+   *
+   * @param createIfNotExist Set this to true if the caller is a write operation
+   *                         in order to create the volume and bucket.
+   * @throws IOException Exceptions other than OMException with result code
+   *                     VOLUME_NOT_FOUND or BUCKET_NOT_FOUND.
+   */
+  private OzoneBucket getBucket(String volumeStr, String bucketStr,
+      boolean createIfNotExist) throws IOException {
+
+    OzoneBucket bucket;
+    try {
+      bucket = proxy.getBucketDetails(volumeStr, bucketStr);
+    } catch (OMException ex) {
+      if (createIfNotExist) {
+        // Note: getBucketDetails always throws BUCKET_NOT_FOUND, even if
+        // the volume doesn't exist.
+        if (ex.getResult().equals(BUCKET_NOT_FOUND)) {
+          OzoneVolume volume;
+          try {
+            volume = proxy.getVolumeDetails(volumeStr);
```

Comment:
should we do getVolumeDetails/create first and if that fails, there is no need to create bucket.
As createVolume is an admin only operation, I would expect most of the client driven operation will fail here.

Reply:
@xiaoyuyao Actually this part was coded previously that it checks if the volume exists first. Then @bharatviswa504 argued that this way we would always have at least two round trips for each `getBucket()` (first `getVolumeDetails`, then `getBucketDetails`) regardless of whether the bucket already exists. This change is made in d2a01d19f0cfa2cdce5e252be2dd33fcb0d4fb6a.